### PR TITLE
Fix wmma api parity

### PR DESCRIPTION
**Pull request by Lzy17**
hipify the wmma api call with rocwmma

### CODE REVIEW

File: csrc/kernels.hip [hunk 1]
```diff
@@ -3093,10 +3093,10 @@ template <typename T, int BITS, int THREADS> __global__ void gemm_device(int M,
   __shared__ T smem_B[2*batch_size_warps*16*32 + (2*16*(batch_size_warps-1))];
   //__shared__ T smem_C[8*32];
 
-   wmma::fragment<wmma::matrix_a, 8, 32, 16, half, wmma::row_major> a_frag;
-   wmma::fragment<wmma::matrix_b, 8, 32, 16, half, wmma::col_major> b_frag;
-   wmma::fragment<wmma::accumulator, 8, 32, 16, half> c_frag;
-   wmma::fill_fragment(c_frag, 0.0f);
+   rocwmma::fragment<wmma::matrix_a, 8, 32, 16, half, wmma::row_major> a_frag;
```
Comment (pnunna93) [inline csrc/kernels.hip]: Replace matrix_a and row_major with rocwmma equivalents

**Review by pnunna93 — CHANGES_REQUESTED** [review]
Please change datatypes and enums also to rocwmma equivalents. I have commented on one of them above.

File: csrc/kernels.hip (hunk 1 above)
Comment (Lzy17) [reply csrc/kernels.hip]: Fixed!

**Comment by seungduk-yanolja** [comment]
Hello, I am sorry to disturb you but I cannot find any place I can report an issue related to this library.
I am testing the new MI300X machine and having trouble importing `bitsandbytes` after successful installation.
I am unsure if I can disclose the error message here, so I am waiting for a response from any developers on the AMD's side.
Please reach out to me through seungduk.kim@yanolja.com
I am a software engineer and I can explain what I am experiencing now.
Thanks!

**Comment by pnunna93** [comment]
> Hello, I am sorry to disturb you but I cannot find any place I can report an issue related to this library. I am testing the new MI300X machine and having trouble importing `bitsandbytes` after successful installation. I am unsure if I can disclose the error message here, so I am waiting for a response from any developers on the AMD's side. Please reach out to me through [seungduk.kim@yanolja.com](mailto:seungduk.kim@yanolja.com) I am a software engineer and I can explain what I am experiencing now. Thanks!

Hi @seungduk-yanolja , please try installing it from [rocm_enabled](https://github.com/ROCm/bitsandbytes/tree/rocm_enabled) branch, the instructions are on that page, please be aware that full enablement is still pending. You can report any future issues on https://github.com/ROCm/rocm repo.

**Comment by seungduk-yanolja** [comment]
> > Hello, I am sorry to disturb you but I cannot find any place I can report an issue related to this library. I am testing the new MI300X machine and having trouble importing `bitsandbytes` after successful installation. I am unsure if I can disclose the error message here, so I am waiting for a response from any developers on the AMD's side. Please reach out to me through [seungduk.kim@yanolja.com](mailto:seungduk.kim@yanolja.com) I am a software engineer and I can explain what I am experiencing now. Thanks!
> 
> Hi @seungduk-yanolja , please try installing it from [rocm_enabled](https://github.com/ROCm/bitsandbytes/tree/rocm_enabled) branch, the instructions are on that page, please be aware that full enablement is still pending. You can report any future issues on https://github.com/ROCm/rocm repo.

Reported the issue here: https://github.com/ROCm/ROCm/issues/2885

Hi @pnunna93, yes, I installed it from the `rocm_enabled` branch because I saw PRs were merged into this branch.
ChatGPT said this:

The backtrace provided indicates that the core dump resulted from a segmentation fault (SIGABRT) triggered within the Python process. Specifically, the crash occurs during the dynamic loading of a shared library related to the `torch` package, more precisely within `libhipblaslt.so`, which is part of the ROCm platform for AMD GPUs. This suggests an issue related to the HIP/ROCm ecosystem, possibly due to an incompatibility or a bug in the library or its dependencies.

The key points in the backtrace indicating the source of the issue are:

- The termination happens after an attempt to load `ExtOpMasterLibrary` from `libhipblaslt.so`, which is part of the ROCm software stack.
- The crash is preceded by a `std::runtime_error`, indicating that an exception was thrown within the C++ standard library, leading to a call to `std::terminate()`, which then causes the process to abort.

Given the complexity of debugging segmentation faults in dynamically loaded libraries, especially within the context of GPU computing, resolving such issues can sometimes require deep technical knowledge of the libraries and the underlying hardware. Collaboration with the community or seeking support from the developers of the libraries involved may be necessary.

**Comment by pnunna93** [comment]
> > > Hello, I am sorry to disturb you but I cannot find any place I can report an issue related to this library. I am testing the new MI300X machine and having trouble importing `bitsandbytes` after successful installation. I am unsure if I can disclose the error message here, so I am waiting for a response from any developers on the AMD's side. Please reach out to me through [seungduk.kim@yanolja.com](mailto:seungduk.kim@yanolja.com) I am a software engineer and I can explain what I am experiencing now. Thanks!
> > 
> > 
> > Hi @seungduk-yanolja , please try installing it from [rocm_enabled](https://github.com/ROCm/bitsandbytes/tree/rocm_enabled) branch, the instructions are on that page, please be aware that full enablement is still pending. You can report any future issues on https://github.com/ROCm/rocm repo.
> 
> Reported the issue here: [ROCm/ROCm#2885](https://github.com/ROCm/ROCm/issues/2885)
> 
> Hi @pnunna93, yes, I installed it from the `rocm_enabled` branch because I saw PRs were merged into this branch. ChatGPT said this:
> 
> The backtrace provided indicates that the core dump resulted from a segmentation fault (SIGABRT) triggered within the Python process. Specifically, the crash occurs during the dynamic loading of a shared library related to the `torch` package, more precisely within `libhipblaslt.so`, which is part of the ROCm platform for AMD GPUs. This suggests an issue related to the HIP/ROCm ecosystem, possibly due to an incompatibility or a bug in the library or its dependencies.
> 
> The key points in the backtrace indicating the source of the issue are:
> 
> * The termination happens after an attempt to load `ExtOpMasterLibrary` from `libhipblaslt.so`, which is part of the ROCm software stack.
> * The crash is preceded by a `std::runtime_error`, indicating that an exception was thrown within the C++ standard library, leading to a call to `std::terminate()`, which then causes the process to abort.
> 
> Given the complexity of debugging segmentation faults in dynamically loaded libraries, especially within the context of GPU computing, resolving such issues can sometimes require deep technical knowledge of the libraries and the underlying hardware. Collaboration with the community or seeking support from the developers of the libraries involved may be necessary.

Hi @seungduk-yanolja , please reinstall hipblaslt with these steps:
git clone --recurse https://github.com/ROCmSoftwarePlatform/hipBLASLt
cd hipBLASLt
git checkout 4b3b34405e7e25cff404f69bfd0a832644430477
./install.sh -idc

You may need to copy and relink hipblaslt .so files from build dir to /opt/rocm/lib/ if it doesn't automatically get replaced after build.

**Comment by seungduk-yanolja** [comment]
> > > > Hello, I am sorry to disturb you but I cannot find any place I can report an issue related to this library. I am testing the new MI300X machine and having trouble importing `bitsandbytes` after successful installation. I am unsure if I can disclose the error message here, so I am waiting for a response from any developers on the AMD's side. Please reach out to me through [seungduk.kim@yanolja.com](mailto:seungduk.kim@yanolja.com) I am a software engineer and I can explain what I am experiencing now. Thanks!
> > > 
> > > 
> > > Hi @seungduk-yanolja , please try installing it from [rocm_enabled](https://github.com/ROCm/bitsandbytes/tree/rocm_enabled) branch, the instructions are on that page, please be aware that full enablement is still pending. You can report any future issues on https://github.com/ROCm/rocm repo.
> > 
> > 
> > Reported the issue here: [ROCm/ROCm#2885](https://github.com/ROCm/ROCm/issues/2885)
> > Hi @pnunna93, yes, I installed it from the `rocm_enabled` branch because I saw PRs were merged into this branch. ChatGPT said this:
> > The backtrace provided indicates that the core dump resulted from a segmentation fault (SIGABRT) triggered within the Python process. Specifically, the crash occurs during the dynamic loading of a shared library related to the `torch` package, more precisely within `libhipblaslt.so`, which is part of the ROCm platform for AMD GPUs. This suggests an issue related to the HIP/ROCm ecosystem, possibly due to an incompatibility or a bug in the library or its dependencies.
> > The key points in the backtrace indicating the source of the issue are:
> > 
> > * The termination happens after an attempt to load `ExtOpMasterLibrary` from `libhipblaslt.so`, which is part of the ROCm software stack.
> > * The crash is preceded by a `std::runtime_error`, indicating that an exception was thrown within the C++ standard library, leading to a call to `std::terminate()`, which then causes the process to abort.
> > 
> > Given the complexity of debugging segmentation faults in dynamically loaded libraries, especially within the context of GPU computing, resolving such issues can sometimes require deep technical knowledge of the libraries and the underlying hardware. Collaboration with the community or seeking support from the developers of the libraries involved may be necessary.
> 
> Hi @seungduk-yanolja , please reinstall hipblaslt with these steps: git clone --recurse https://github.com/ROCmSoftwarePlatform/hipBLASLt cd hipBLASLt git checkout 4b3b34405e7e25cff404f69bfd0a832644430477 ./install.sh -idc
> 
> You may need to copy and relink hipblaslt .so files from build dir to /opt/rocm/lib/ if it doesn't automatically get replaced after build.

It looks like the same command lines as described in the `README.md` of `rocm_enabled` branch. I used these but let me retry.

Update: I tried to install hipBLASLt again but there was an error (invalid memory access) and the whole filesystem became read-only. I rebooted the machine and then it did not correctly recognize the GPUs. I rebooted the IPMI and then it became normal. At this moment, what I can do with this machine (MI300X) is run vLLM with 4 out of 8 GPUs because the output became so weird when I used all 8 GPUs. Will try and explore more what I can do.

**Comment by Titus-von-Koeller** [comment]
Hey all! I'm Titus, one of the bitsandbytes maintainers. We currently have a strong push underway to officially make different hardware backends than CUDA possible in BNB. Would you be willing to help us to get the AMD part right and consolidate the code-bases?

**Comment by amathews-amd** [comment]
> Hey all! I'm Titus, one of the bitsandbytes maintainers. We currently have a strong push underway to officially make different hardware backends than CUDA possible in BNB. Would you be willing to help us to get the AMD part right and consolidate the code-bases?

Hi @Titus-von-Koeller , sure! we were planning to reach out to you once we closed some internal dependencies.  Is there a forum we can discuss ?

**Comment by pnunna93** [comment]
> It looks like the same command lines as described in the `README.md` of `rocm_enabled` branch. I used these but let me retry.
> 
> Update: I tried to install hipBLASLt again but there was an error (invalid memory access) and the whole filesystem became read-only. I rebooted the machine and then it did not correctly recognize the GPUs. I rebooted the IPMI and then it became normal. At this moment, what I can do with this machine (MI300X) is run vLLM with 4 out of 8 GPUs because the output became so weird when I used all 8 GPUs. Will try and explore more what I can do.

Hi @seungduk-yanolja, sounds like there is an issue with hipblaslt build/linking. The version I pointed to has ExtOpMasterLibrary class but something else is going wrong in the build. Please check back on the ROCm issue, they would be able to help. Thanks.

**Comment by seungduk-yanolja** [comment]
Thank you all. I do not have access to the machine anymore since it was a short-time PoC. There is another PoC scheduled next month so will try again. Thanks again.

**Pull request merged**